Sources/ClientRuntime/Networking/Http/URI.swift

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+import Foundation
+
+public struct URI: Hashable {
+    public let scheme: Scheme
+    public let path: String
+    public let host: String
+    public let port: Int16
+    public let query: [SDKURLQueryItem]
+    public let username: String?
+    public let password: String?
+    public var url: URL? {
+        self.toBuilder().getUrl()
+    }
+    public var queryString: String? {
+        if self.query.isEmpty {
+            return nil
+        }
+
+        return self.query.map { queryItem in
+            return [queryItem.name, queryItem.value].compactMap { $0 }.joined(separator: "=")
+        }.joined(separator: "&")
```

Comment:
I think I saw this logic somewhere else too, you might be able to extract it into an extension method/property on `Array<SDKURLQueryItem>`